strawberryfields/apps/train/parametrization.py

```diff
@@ -0,0 +1,312 @@
+    """
+
+    def __init__(
+        self,
+        A: np.ndarray,
+        n_mean: float,
+        embedding,
+        threshold: bool,
+        samples: Optional[np.ndarray] = None,
+    ):
+        if not np.allclose(A, A.T):
+            raise ValueError("Input must be a NumPy array corresponding to a symmetric matrix")
+        self.A_init = rescale_adjacency(A, n_mean, threshold)
+        self._A_init_samples = None
+        self.embedding = embedding
+        self.threshold = threshold
+        self.n_modes = len(A)
+        if samples:
+            self.add_A_init_samples(samples)
+
+    def W(self, params: np.ndarray) -> np.ndarray:
+        r"""Calculate the diagonal matrix of weights :math:`W` that depends on the trainable
+        parameters :math:`\theta`.
+
+        **Example usage:**
+
+        >>> vgbs.W(params)
+        array([[0.97530991, 0.        , 0.        , 0.        ],
+               [0.        , 0.95122942, 0.        , 0.        ],
+               [0.        , 0.        , 0.99004983, 0.        ],
+               [0.        , 0.        , 0.        , 0.99501248]])
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            array: the diagonal matrix of weights
+        """
+        return np.sqrt(np.diag(self.embedding(params)))
+
+    def A(self, params: np.ndarray) -> np.ndarray:
+        r"""Calculate the trained adjacency matrix :math:`A(\theta)`.
+
+        **Example usage:**
+
+        >>> vgbs.A(params)
+        array([[0.        , 0.30298161, 0.31534653, 0.31692721],
+               [0.30298161, 0.        , 0.30756059, 0.30910225],
+               [0.31534653, 0.30756059, 0.        , 0.32171695],
+               [0.31692721, 0.30910225, 0.32171695, 0.        ]])
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            array: the trained adjacency matrix
+        """
+        return self.W(params) @ self.A_init @ self.W(params)
+
+    def generate_samples(self, A: np.ndarray, n_samples: int, **kwargs) -> np.ndarray:
+        """Generate GBS samples from an input adjacency matrix.
+
+        **Example usage:**
+
+        >>> vgbs.generate_samples(vgbs.A(params), 2)
+        array([[1, 1, 1, 0],
+               [0, 1, 0, 1]])
+
+        Args:
+            A (array): the adjacency matrix
+            n_samples (int): the number of GBS samples to generate
+            **kwargs: additional arguments to pass to the sampler from
+                `The Walrus <https://the-walrus.readthedocs.io/en/stable/>`__
+
+        Returns:
+            array:  the generated samples
+        """
+        cov = A_to_cov(A)
+
+        if self.threshold:
+            samples = thewalrus.samples.torontonian_sample_state(
+                cov, n_samples, hbar=sf.hbar, **kwargs
+            )
+        else:
+            samples = thewalrus.samples.hafnian_sample_state(cov, n_samples, hbar=sf.hbar, **kwargs)
```

Comment:
@nquesada we are using `hbar=sf.hbar`, do you think that will work ok (tests are passing!)